### ISSUE TITLE
when does `wire.get` work at `get: wire.get || adapter.read`?

### ISSUE THREAD

**Issue by swhgoon** [PsychoLlama/gun-level#15]
I am wondering about this code... as a `gun` newbie...:)
```
// Register the driver.
gun.opt({
		wire: {
			get: wire.get || adapter.read,
			put: wire.put || adapter.write,
		},
}, true);
```

**Comment by PsychoLlama** [comment]
Hey @swhgoon! Thanks for the question :)

**TL;DR:**
The code is run whenever a gun instance is created, or when any `gun.opt` is called without passing `true` as the second param. `wire.get` may or may not be defined.

**More details:**

When you first instantiate gun, you can pass an options object:

```js
Gun({
  peers: ['http://url.com'],
  init: true,

  wire: { /* driver hooks */ },
})
```

One of the options you can pass is `wire`, an object containing hooks for reads and writes.
There are two hooks, `get` and `put`. They're called by `gun.get` and `gun.put` after gun runs inputs through some validation.

Whenever options are added or changed on a gun instance, it fires a global `opt` event with the gun instance and the new options (which is why the code snipped you've included is surrounded by `Gun.on('opt').event(function....`). This provides a hook for drivers to mount when a gun instance is created, or to unmount when options change.

There's a method on each gun instance that will change the options, called `.opt` which triggers the `opt` event again. Since `.opt` is called from within the `opt` event, that would create an infinite loop. Passing `true` as a second argument prevents the `opt` event from firing, preventing that loop.

Since `opt` may fire multiple times, gun-level might've already mounted itself onto the gun instance, so we check to see if `wire.get` or `wire.put` already exists, otherwise we register them.

One of the nuances to the plugin system is how other modules (like the json driver or websockets) coexist, and I'm a bit fuzzy on those details. It seems like it should only allow one driver at a time, but wsp.js, file.js, and gun-level can still work together without knowing about each other. It seems a bit magical, but I haven't looked into it (maybe @amark could help clarify?).

So to summarize, when you call `Gun()`, it fires the global `opt` event, and gun-level mounts itself onto the gun instance using `gun.opt`. Since the event may fire many times, gun-level needs to make sure it hasn't already mounted.

Phew! Words. Hopefully that answers your question :)

**Comment by amark** [comment]
@swhgoon in the first iterations of GUN we didn't worry about the plugin architecture. That quickly has caught up to us, and how it works (which isn't very good) is what @PsychoLlama was explaining above. We're changing the plugin architecture to something that should at least be decent with the upcoming `0.5` version of GUN.

**Comment by amark** [comment]
@swhgoon and thanks for jumping into the community and trying to participate and contribute! :D
